### PR TITLE
implement ed25519 support

### PR DESCRIPTION
**Pull request by whyrusleeping**

### CODE REVIEW

**Comment by whyrusleeping** [comment]
cc @lgierth @Kubuxu for review

File: ed25519_test.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+	ok, err := pub.Verify(data, sig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !ok {
+		t.Fatal("signature didnt match")
+	}
+}
```
Comment (Kubuxu) [inline ed25519_test.go]: Other basic test case is that signatures of wrong key shouldn't match.

File: key.go [hunk 2]
```diff
@@ -59,18 +60,13 @@ type PrivKey interface {
...
 }
 
 type PubKey interface {
 	Key
 
 	// Verify that 'sig' is the signed hash of 'data'
 	Verify(data []byte, sig []byte) (bool, error)
-
-	// Encrypt data in a way that can be decrypted by a paired private key
-	Encrypt(data []byte) ([]byte, error)
```
Comment (Kubuxu) [inline key.go]: You don't have remove those methods, ed25519 can be transposed into cure25519 keys, and they don't leak information like RSA.

Comment (whyrusleeping) [inline key.go]: oh? iiiinteresting

Comment (whyrusleeping) [inline key.go]: Do you know where to corrolary would be in the package? github.com/agl/ed25519

Comment (Kubuxu) [inline key.go]: Yup, NaCL implements it also, it is supposed to be also proven to be secure http://crypto.stackexchange.com/questions/3260/using-same-keypair-for-diffie-hellman-and-signing/3311#3311 which isn't true for RSA.

**Comment by Kubuxu** [comment]
ed25519 was not vendored into gx.

File: key.go (hunk 2 above)
Comment (whyrusleeping) [inline key.go]: @Kubuxu is cryptobox what we want? https://godoc.org/golang.org/x/crypto/nacl/box#Seal

Comment (Kubuxu) [inline key.go]: We use exactly the same primitives in cjdns but this doesn't include key conversion and we might want to use those primitives in our own way.

Comment (Kubuxu) [inline key.go]: Here is key conversion: http://crypto.stackexchange.com/questions/9536/converting-ed25519-public-key-to-a-curve25519-public-key

**Comment by Kubuxu** [comment]
It might be too much, but it would be great to create tests for official test vectors of ed25519 so if we change the underlying impl we are sure it will still work: https://tools.ietf.org/html/draft-josefsson-eddsa-ed25519-02#section-6

**Comment by vyzo** [comment]
Guys, what's the status of this PR and what kind of help do you need?

**Comment by whyrusleeping** [comment]
@vyzo we got distracted after trying to find out how to turn ed25519 keys into curve25519 keys for encrypting things. If you can figure out how that works, that would be pretty helpful

**Comment by vyzo** [comment]
Is encryption actually necessary from a protocol point of view?
Or is it something nice to have for completeness?

**Comment by vyzo** [comment]
Also, is there any reason for not using the [golang.org/crypto/ed25519](https://github.com/golang/crypto/tree/master/ed25519) package?

**Comment by lgierth** [comment]
> Also, is there any reason for not using the golang.org/crypto/ed25519 package?

Yeah the golang one [1] seems to be the same continuation of agl's [2]

[1] https://github.com/golang/crypto/tree/master/ed25519
[2] https://github.com/agl/ed25519

**Comment by vyzo** [comment]
Right, the package graduated from agl's personal github to golang.org.

**Comment by vyzo** [comment]
So digging int the agl code I found this file:
https://github.com/agl/ed25519/blob/master/extra25519/extra25519.go
which provides conversion functions from ed25519 keys to curve25519 keys.

Is this what you were looking for?

Note: sadly the extra25519 file is not included in the golang.org package.

**Comment by whyrusleeping** [comment]
@vyzo No, encryption isnt required as part of the interface. I don't believe we use it anywhere, but it would be a nice to have down the road. We can shrink the interface for now, and when we get to the point where we need it we can finish this stuff

**Comment by vyzo** [comment]
@whyrusleeping shrinking the interface for quick merging sounds great, we really want ECC keys!

**Comment by parkan** [comment]
Anything we can do to help get this merged in? We're blocked by this and happy to help out

**Comment by lgierth** [comment]
- [ ] use ed25519 from stdlib
- [ ] it's a bit hard to review this without any indication of what the scope is -- is this supposed to let us do `ipfs swarm connect /ip4/1.2.3.4/tcp/4001/ipfs/<ed25519key>`? or ed25519 signatures for IPNS?

generally i'm all for it! :) support for ed25519 keys is great because it'll also enable SHS or CryptoAuth transport-level crypto.

**Comment by lgierth** [comment]
>  it's a bit hard to review this without any indication of what the scope is -- is this supposed to let us do `ipfs swarm connect /ip4/1.2.3.4/tcp/4001/ipfs/<ed25519key>`? or ed25519 signatures for IPNS?

lol nevermind, I see what the scope is for now :)

**Comment by whyrusleeping** [comment]
@parkan right now the hard part is the migration path. How do we best ship this in a way that makes sense.

My current thoughts are to ship this into go-ipfs, so that it enables nodes to be able to handle these keys. And then only allow generation of these keys with a special flag on init.

The issue is that if your node is using an ed25519 key, you won't be able to communicate with a node who doesnt know how to handle that key.

**Comment by vyzo** [comment]
@whyrusleeping Can't it be merged into libp2p regardless of IPFS migration?

**Comment by whyrusleeping** [comment]
@vyzo Yeap! and we're gonna do that now! (Sorry about the delay, chinese internet is super difficult)

**Pull request merged**